### PR TITLE
Update reference list nodes

### PR DESCRIPTION
**Pull request by atothhpe**
Had to add Entity.LIST_NODE, see [Common PR](https://github.com/MicroFocus/octane-plugin-common/pull/24)
Added proper support for selection handlers on EntityComboBox
Added Reference field editor, atm just contains an EntityComboBox
Created entity loader for list nodes based on metadata

### CODE REVIEW

File: octane-eclipse-plugin/src/com/hpe/octane/ideplugins/eclipse/ui/entitydetail/field/FieldEditorFactory.java [hunk 1]
```diff
@@ -58,19 +91,27 @@ public FieldEditor createFieldEditor(Composite parent, EntityModelWrapper entity
...
+
         try {
+
             fieldEditor.setField(entityModelWrapper, fieldName);
+
         } catch (Exception ex) {
-            ILog log = Activator.getDefault().getLog();
             StringBuilder sbMessage = new StringBuilder();
-            sbMessage.append("Faied to set field ")
+            sbMessage.append("Faied to set field  ")
```
Comment (paul-ciobanu) [inline octane-eclipse-plugin/src/com/hpe/octane/ideplugins/eclipse/ui/entitydetail/field/FieldEditorFactory.java]: @atothhpe Why did you add an empty space? And since you touched this can you change to "Failed?

File: octane-eclipse-plugin/src/com/hpe/octane/ideplugins/eclipse/ui/entitydetail/field/FieldEditorFactory.java [hunk 2]
```diff
@@ -12,25 +12,58 @@
...
 
 public class FieldEditorFactory {
 
+    private static final int COMBO_BOX_ENTITY_LIMIT = 100;
+
+    private static final LabelProvider DEFAULT_ENTITY_LABEL_PROVIDER = new LabelProvider() {
+        @Override
+        public String getText(Object element) {
+
+            EntityModel entityModel = (EntityModel) element;
```
Comment (paul-ciobanu) [inline octane-eclipse-plugin/src/com/hpe/octane/ideplugins/eclipse/ui/entitydetail/field/FieldEditorFactory.java]: @atothhpe Can element be null or of type different than EntityModel? Would it make sense to be proactive here and check for these cases?

File: octane-eclipse-plugin/src/com/hpe/octane/ideplugins/eclipse/ui/entitydetail/field/FieldEditorFactory.java [hunk 3]
```diff
@@ -12,25 +12,58 @@
...
+
+            if (Entity.getEntityType(entityModel) == Entity.WORKSPACE_USER) {
+                return Util.getUiDataFromModel(entityModel.getValue(EntityFieldsConstants.FIELD_FULL_NAME));
+
+            } else {
+                return Util.getUiDataFromModel(entityModel.getValue(EntityFieldsConstants.FIELD_NAME));
```
Comment (paul-ciobanu) [inline octane-eclipse-plugin/src/com/hpe/octane/ideplugins/eclipse/ui/entitydetail/field/FieldEditorFactory.java]: @atothhpe Minor feedback: instead of having 2 return paths and since the code is similar, I would have used the if to decide on the field name and then have only one return statement. Up to you

File: octane-eclipse-plugin/src/com/hpe/octane/ideplugins/eclipse/ui/entitydetail/field/FieldEditorFactory.java [hunk 4]
```diff
@@ -85,4 +126,67 @@ public FieldEditor createFieldEditor(Composite parent, EntityModelWrapper entity
...
+
+        ReferenceFieldEditor fieldEditor = new ReferenceFieldEditor(parent, SWT.NONE);
+
+        if (fieldMetadata.getFieldTypedata().isMultiple()) {
+            fieldEditor.setSelectionMode(SWT.MULTI);
+        } else {
+            fieldEditor.setSelectionMode(SWT.SINGLE);
+        }
+
+        Target traget = targets[0];
```
Comment (paul-ciobanu) [inline octane-eclipse-plugin/src/com/hpe/octane/ideplugins/eclipse/ui/entitydetail/field/FieldEditorFactory.java]: @atothhpe "target"

File: octane-eclipse-plugin/src/com/hpe/octane/ideplugins/eclipse/ui/entitydetail/field/FieldEditorFactory.java [hunk 5]
```diff
@@ -85,4 +126,67 @@ public FieldEditor createFieldEditor(Composite parent, EntityModelWrapper entity
...
+                        entities
+                        .stream()
+                        .filter(entityModel -> stringLike(Util.getUiDataFromModel(entityModel.getValue(EntityFieldsConstants.FIELD_NAME)), searchQuery))
+                        .collect(Collectors.toList());
+                }
+                
+                return entities;
+            });
+        } else {
+            disposeFieldEditor(fieldEditor);
```
Comment (paul-ciobanu) [inline octane-eclipse-plugin/src/com/hpe/octane/ideplugins/eclipse/ui/entitydetail/field/FieldEditorFactory.java]: @atothhpe Is there a point in instantiating a new fieldEditor if the target type != LIST_NODE? Perhaps it can be refactored:
if (target type not supported)
  throw
else {
  instantiate fieldEditor
  fieldEditor.setSelectionMode
  fieldEditor.setEntityLoaded...
}

File: octane-eclipse-plugin/src/com/hpe/octane/ideplugins/eclipse/ui/entitydetail/field/FieldEditorFactory.java [hunk 6]
```diff
@@ -85,4 +126,67 @@ public FieldEditor createFieldEditor(Composite parent, EntityModelWrapper entity
...
+                        Query.statement("logical_name", QueryMethod.EqualTo, logicalName));
+                                
+                Collection<EntityModel> entities = entityService.findEntities(Entity.LIST_NODE, qb, null);
+                
+                //for some reason list nodes are not server side filterable, so you have to do it client side   
+                if(!searchQuery.isEmpty()) {
+                    entities =
+                        entities
+                        .stream()
+                        .filter(entityModel -> stringLike(Util.getUiDataFromModel(entityModel.getValue(EntityFieldsConstants.FIELD_NAME)), searchQuery))
```
Comment (paul-ciobanu) [inline octane-eclipse-plugin/src/com/hpe/octane/ideplugins/eclipse/ui/entitydetail/field/FieldEditorFactory.java]: @atothhpe Perhaps searchQuery can be trimmed and toLower only once? That might imply also changing stringLike

File: octane-eclipse-plugin/src/com/hpe/octane/ideplugins/eclipse/ui/util/EntityComboBox.java [hunk 7]
```diff
@@ -133,73 +133,115 @@ public EntityComboBox(Composite parent, int style, LabelProvider labelProvider,
...
+                            Label lblNoResult = new Label(rootComposite, SWT.NONE);
+                            lblNoResult.setText("No results");
+                            lblNoResult.setLayoutData(new GridData(SWT.CENTER, SWT.CENTER, true, true));
+                        } else if (selectionMode == SWT.SINGLE) {
+                            showSingleSelectionUI();
+                        } else if (selectionMode == SWT.MULTI) {
+                            showMultiSelectionUI();
+                        }
+
+                    } else if (Status.WARNING == status.getSeverity()) {
```
Comment (paul-ciobanu) [inline octane-eclipse-plugin/src/com/hpe/octane/ideplugins/eclipse/ui/util/EntityComboBox.java]: @atothhpe Is it ok to use getSeverity() and above just status? What about the cancel or error cases?

File: octane-eclipse-plugin/src/com/hpe/octane/ideplugins/eclipse/ui/entitydetail/field/FieldEditorFactory.java (hunk 1 above)
Comment (atothhpe) [reply octane-eclipse-plugin/src/com/hpe/octane/ideplugins/eclipse/ui/entitydetail/field/FieldEditorFactory.java]: accident, fixing

File: octane-eclipse-plugin/src/com/hpe/octane/ideplugins/eclipse/ui/entitydetail/field/FieldEditorFactory.java (hunk 2 above)
Comment (atothhpe) [reply octane-eclipse-plugin/src/com/hpe/octane/ideplugins/eclipse/ui/entitydetail/field/FieldEditorFactory.java]: I can typecheck, but it would be impossible for an element in an EntityComboBox to be null or another type, if you tried using the label provider anywhere else, it can happen.

File: octane-eclipse-plugin/src/com/hpe/octane/ideplugins/eclipse/ui/entitydetail/field/FieldEditorFactory.java (hunk 4 above)
Comment (atothhpe) [reply octane-eclipse-plugin/src/com/hpe/octane/ideplugins/eclipse/ui/entitydetail/field/FieldEditorFactory.java]: thanks

File: octane-eclipse-plugin/src/com/hpe/octane/ideplugins/eclipse/ui/entitydetail/field/FieldEditorFactory.java (hunk 5 above)
Comment (atothhpe) [reply octane-eclipse-plugin/src/com/hpe/octane/ideplugins/eclipse/ui/entitydetail/field/FieldEditorFactory.java]: that's a great idea, I forgot I could just save the loader in a local variable, (dum dum)

File: octane-eclipse-plugin/src/com/hpe/octane/ideplugins/eclipse/ui/entitydetail/field/FieldEditorFactory.java (hunk 6 above)
Comment (atothhpe) [reply octane-eclipse-plugin/src/com/hpe/octane/ideplugins/eclipse/ui/entitydetail/field/FieldEditorFactory.java]: I'm not sure where it's being done twice atm

File: octane-eclipse-plugin/src/com/hpe/octane/ideplugins/eclipse/ui/util/EntityComboBox.java (hunk 7 above)
Comment (atothhpe) [reply octane-eclipse-plugin/src/com/hpe/octane/ideplugins/eclipse/ui/util/EntityComboBox.java]: Cancel and error cases can't happen unless you return them explicitly in the job method
